File: 序号/46.全排列.kt
import java.util.*

class Solution {
    val ans = mutableListOf<MutableList<Int>>()
    /** 方法一  list.contains 为线性搜索，可以优化，其实就是空间换时间
    fun permute(nums: IntArray): List<List<Int>> {
        //全排列，定下左边的，尝试遍历右边的，尝试完之后回来
        backTrack(LinkedList(),nums)
        return ans
    }

    fun backTrack(list: LinkedList<Int>, nums: IntArray) {
        if (list.size == nums.size) {
            ans.add(LinkedList(list))
            return
        }
        for (i in nums) {
            if (list.contains(i)) continue
            list.addLast(i)
            backTrack(list, nums)
            list.removeLast()
        }
    }*/


    private lateinit var used:BooleanArray
    fun permute(nums: IntArray): List<List<Int>> {
        //全排列，定下左边的，尝试遍历右边的，尝试完之后回来
        used = BooleanArray(nums.size)
        backTrack(LinkedList(),nums)
        return ans
    }

    fun backTrack(list: LinkedList<Int>, nums: IntArray) {
        if (list.size == nums.size) {
            ans.add(LinkedList(list))
            return
        }
        for (i in nums.indices) {
            if (used[i]) continue
            list.addLast(nums[i])
            used[i]=true
            backTrack(list, nums)
            list.removeLast()
            used[i]=false
        }
    }

}

File: 序号/46.全排列.cpp
class Solution {
 public:
  vector<vector<int>> ans;
  vector<int> tmp;
  vector<int>* used;
  vector<int>* nums;
  int n;
  void backTrace() {
    if (tmp.size() == n) {
      ans.emplace_back(tmp);
      return;
    }
    for (int i = 0; i < n; i++) {
      if(used->at(i)==1) continue;
      tmp.emplace_back(nums->at(i));
      used->at(i)=1;
      backTrace();
      tmp.pop_back();
      used->at(i)=0;
    }
  }
  vector<vector<int>> permute(vector<int>& _nums) {
    used = new vector<int>(_nums.size(), 0);
    nums = &_nums;
    n=nums->size();
    backTrace();
    delete used;
    return ans;
  }
};